File: backend-go/internal/domain/engine/option_choices.go
package engine

import (
	"strings"

	"github.com/tiagofur/muebles-backend/internal/domain"
)

// EffectiveOptionChoices merges project-level defaults with item overrides (F029).
// Empty item values inherit the project default; explicit item values win.
func EffectiveOptionChoices(item map[string]string, projectLevel map[string]string) map[string]string {
	out := make(map[string]string)
	for code, id := range projectLevel {
		v := strings.TrimSpace(id)
		if v != "" {
			out[code] = v
		}
	}
	for code, id := range item {
		v := strings.TrimSpace(id)
		if v != "" {
			out[code] = v
		}
	}
	return out
}

// choicesForItem is a convenience wrapper for project item resolution.
func choicesForItem(project domain.Project, item domain.ProjectItem) map[string]string {
	return EffectiveOptionChoices(item.OptionChoices, project.ProjectLevelChoices)
}

File: packages/domain/src/optionChoices.ts
/**
 * Effective option resolution: project defaults + per-line overrides (F029 / #35).
 */

import type { OptionChoices } from './types';

/**
 * Merge project-level defaults with item overrides.
 * Item keys with empty/whitespace values are ignored (inherit project).
 * Explicit item values win over project defaults.
 */
export function effectiveOptionChoices(
  itemChoices: OptionChoices | undefined | null,
  projectLevelChoices?: OptionChoices | null,
): OptionChoices {
  const result: Record<string, string> = {};
  if (projectLevelChoices) {
    for (const [code, id] of Object.entries(projectLevelChoices)) {
      const v = id?.trim();
      if (v) result[code] = v;
    }
  }
  if (itemChoices) {
    for (const [code, id] of Object.entries(itemChoices)) {
      const v = id?.trim();
      if (v) result[code] = v;
    }
  }
  return result;
}
